{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#global params.\n",
    "\n",
    "#game title. full list of games = http://yavar.naddaf.name/ale/list_of_current_games.html\n",
    "GAME=\"ms_pacman\"\n",
    "\n",
    "#game image will be resized from (210,160) to your image_size. \n",
    "#You may want a bigger image for your homework assignment IF you want a larger NN\n",
    "IMAGE_W,IMAGE_H = IMAGE_SIZE =(105,80)\n",
    "\n",
    "#number of parallel agents and batch sequence length (frames)\n",
    "N_AGENTS = 10\n",
    "SEQ_LENGTH = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: THEANO_FLAGS=\"floatX=float32\"\n"
     ]
    }
   ],
   "source": [
    "from __future__ import print_function \n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "%matplotlib inline\n",
    "%env THEANO_FLAGS=\"floatX=float32\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.image.AxesImage at 0x10c8cf0b8>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAASAAAAFjCAYAAACdT9ZCAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAAPYQAAD2EBqD+naQAAIABJREFUeJztnXuUXMV957+/6emZ0UgayRiNcOIHOAL5hRHIhmUNGJuX\nH2DDxmsbx+YkOazjOGRZZ08MrPBaAQw++KxNIJBw1llnjdfYOOuYIDuWBbHDw4s4lhAvExkHYQFC\nDyQ0A/Oento/+nbX7/ZUdd/bfW9X9+j7OUdHNff+7q9+VXWn5le3qn4lxhgQQkgIekIbQAg5dGEH\nRAgJBjsgQkgw2AERQoLBDogQEgx2QISQYLADIoQEgx0QISQY7IAIIcFgB0QICUbQDkhE/kREdojI\nhIg8KCLvDGkPIaS9BOuARORjAP4HgC8COB7AIwA2isjhoWwihLQXCbUZVUQeBLDZGHNp9LMAeBbA\njcaY64MYRQhpK70hMhWRIoC1AK6tXDPGGBG5G8DJDvlXAzgHwDMAJttkJiGkOQYAHAlgozFmfz3B\nIB0QgMMBFADsqbm+B8Bqh/w5AP5P3kYRQjLl9wB8u55AqA4oLc8AwHmrjsa2PXvw3iOPqt6YmJPc\nM79/59M45fVvzDkXOxQeLJT/v2fHjgVa1s7IN9s857cfAOgvHJX26/6yApXy6rJK9Hq+OD6Ou556\nCoh+b+sRqgN6EUAJwMqa6ysB7HbITwLAtj17sH9yAg88t7N6442HDeOYV6/Iy04AQF+hF8OLl+Sa\nh36Bl0St0t9bwBFLbL5jpfw7oPaUtTPyzTbP+e0HxDugSvt1f1mBSnl3vrQPv3xxX+zOVKlUSTb8\nXBKkAzLGzIjIFgBnAPhHoPoR+gwAN/qe+84F5+NLDzyAr3/4d6vXztg0jGcnyunxGfsL2ttjW15/\nZi8Z9y9xQdwf40tGsHf2AO56+YLE8vV067uzyqNZXLQyPztrLwDg4d278Wcnv6t6/fRNw9V0N5TV\nJ+uS2FeTb5o8m5XfO3sAP3rl/NT6XW3oaj8gXtZK+71YOoC7Rm1Zs2w/n3yWZQVsee85z5Z1WW/5\n2sO79+Bd3/ymU3ctIYdgXwXwd1FH9BCAzwEYBPB3AW0ihLSRYB2QMeaOaM3PVSgPvbYBOMcYs6/+\nk4SQhULQj9DGmFsA3JJU/uXpKXxw1SrsH7dDy6cOTlfTs9PKYZz2rLHsn3Nfn6ojXzgWv3lpOrl8\nUt19VrZXpfdPlst31u+swv6JBVJWLevy/Htr8k2TZ7Py5jg8/UIpubyLqN1c7QfEvwFV2m+ueCxe\n0WXNsv188hmWFbDl1e/nXDQse3l6yv28g67bC3b+6mPan+miNe3PE8AHj3atSMiZQGUNkm9fmLL2\nLD6u/ZkGKmsjuq4DIoQsHLplHRAAQETQI4Jij+03i6oLlX7r765bO+LUcc1jS5U+JX9CtvJJZL+k\ndBcLalZL1BhFpTu1rFpey+q5muset7LTaiSgi9pXsE9crvToEVuIMtWTr7ThQmm/evKu97XYY8tX\nSYq4Z9dc0AMihASDHRAhJBhdNQQrO+Nx987oxVR6ysF32eMdSp7yHllNXIc4r3dbWZ0rDlHroruF\nfCJBygQ0bMMF03515DWu65VycwhGCOkK2AERQoIRLCBZGkTkBABbNl54Id4+PIyDah/U2h/a/ax6\nfxSmPW7ggFp8pV3iSd/iLjVlo+V9i8cWlebLTnlsUQvB9F6ibeeqvUSqedZs6LCyuuSTyPpcdL0u\nLk2eWcov8sg3aMOubD9fWX36B6x8pbxbPmgj6iwvlv9/dO9enP3t2wFgrTFmq1tZGXpAhJBgsAMi\nhASjy2bBGqC8yN4BT4gDJaQXXw0NzbrUYCYWssDKFwdmHVeBkZmeeUoKypZYeINWRr9tLGuvKqtm\n1iGfSNY37I+VKXmezcg3bL8aezJvw056VxPqH/ENFVuAHhAhJBjsgAghwVhYQzBFyXdDuf+Daubi\n7rNsGKJKZDcAcDv28Yo7OGtd09M3lqPejatrorv5HCYdvev6Wihr/sFf24+umkbtB7SvDfNoP40u\n64gq06k/tmWFGoL97GyrX5tz8j9Z+VJ0o9X3hB4QISQY7IAIIcFYsEMwTYJJFyzvs0LarZ3xPFvU\nD4sjqZ6bnfAtYmvBl1ePziTQn7asC30I1rD9ah7IvA1zbj+NLqt4yq1Z3OtWlGKLV2LoARFCgsEO\niBASjENiCJaEWeOeTfDPpikZlZ6L0sWC3cfz3058xalCR5ibSzBM1DJJ9F/r0V9KUNaFPgRr1H5A\nujbstPaLoYeSCYbZh/V5gtXnAD0gQkgw2AERQoLBIZiDrCOUeCPSJbGlBf1J6IJoLC2RVfGabcOO\nbj89E6f2kV2lg9Ur8Vk1MusrtJCvgh4QISQY7IAIIcHgECyiV/masUVqniDeuuLUkVDVs5FmZm3f\nvn7zMnemKiKidmljfxUcugFgMqX+fqW/kKSsbo3dTYr2A9K1Yce1X8qyzqo9Yjc+MuS2p2++Pa0O\ncekBEUKCwQ6IEBKMQ3oIFo8O556K8Ic4sPL62eqTnoh3Gr0oTEuMqkz1WVIxLS3oT1LWhTgEi4fj\naNB+QKo67rT206Quq2d/W0sRPD3QAyKEBIMdECEkGAt2CKbXSelY3SX1w7jyWc/ctKKa9gf6thRV\nBDntmU5UNuR4wh4Yj987oWx5z09WwMWkb7NWAv2tlLVXfC75fPlEsj5XXg8DUuTZjHzD9quxp1Ed\nd3L7tVLWPCJ4augBEUKCwQ6IEBKMhTUE02EHfEfp6jOYlMt6cKTolu/3+KlTvuNu5+bbMpmRLYv0\nUb056E9bVpd8ElnfFJtpMs8s5Zus465sv7T2tBLB0wM9IEJIMNgBEUKC0ZVDsFaizOkohNJjH153\n0qhT/hotL1b/ujX15bXuXGzJW3+Ksmp5Lasd9uset7LTJXXelpqh6eu1eV6u9Oj2zsLGxPIp6rjr\n2y+lPZXfO54LRgjpWtgBEUKC0ZVDMI3vrCLfcbfes418Ee8S6G8kn0a2VfnQZU2yiC3J+VJpy5S2\nzlqpg0b2tLO9Q5U1qzPC6AERQoLBDogQEgwxXRCVXEROALBl44UX4u3DwzioQgqs2bCymp6a1uHs\nfIuv1GKt2MI332IwvVhLh7RroD8P3f0e26cDldVlj8+W2EJET/76OKokejqpDjrBlrz191n9A9Hx\n0NvO3VO9tjxaH/no3r04+9u3A8BaY8xWt7Iy9IAIIcFgB0QICUZXzoJph3JZ0Q4hx9SN4iIbvyAe\nTU71uUq+4IlKV1JCetHc0JA7/lxVf0a6/eEWrHxxoH1lbWRP74C7XuLhOBqfY5xIT0fVQRhbNLmX\ndc7+tCT6veu4hYgicoWIPCQioyKyR0T+QUSOcchdJSK7RGRcRDaJyKqsbSGEdDZ5DMFOBXATgJMA\nnAmgCOAnIrKoIiAilwG4BMCnAZwIYAzARhHpy8EeQkiHkvkQzBjzAf2ziPw+gL0A1gK4P7p8KYCr\njTEbIpmLAOwBcD6AO3y65wCUAAz2WnfxX87Zq/K2svqcr4PqzKPTNw5X0+PquuiuWHusSukiNdz7\n6dn7nOKn/LisfyKl7kGl++6zrO5lqqz+oOOWrMqa1p6M1qWlIh5k3hKqDvK0Jc27B+T0/jkWJS5S\noUcr90spJtbb8RF6OcrFPgAAInIUgCMA3FMRMMaMAtgM4OQ22EMI6RBy/QgtIgLgBgD3G2N+GV0+\nAuUOaU+N+J7oXkP0iY76I7TueGMnRoozGQ/CNNE4CJPv47emKtOC7uV99rr+CzTj+cuSd1mT2BPa\nA+qEOmiXLQ3fvRb1p2lvHQKtGfKeBbsFwFsAvCvnfAghXUhuHZCI/BWADwA41Rjzgrq1G+UOdCXi\nXtBKAA/X07n+3nsx1Bf/Tn3+6tW4YPXqTGwmhKTjB9u3487t2wFYr2h0ejrx87l0QFHn82EA7zbG\n7NT3jDE7RGQ3gDMAPBrJD6E8a3ZzPb3rTzsNxw4Pz9tkXXEDY9fVD/qjWCsBzPSzeteAcehPovta\nj+6S57RLr7ubUVlbsSf0EKwT6iBrW5p995Lqz6K9L3A4AI/t3Yv33367M89aMu+AROQWABcC+BCA\nMRGpbNYaMcZMRukbAFwpIr8G8AyAqwE8B+DOrO0hhHQueXhAn0H5b8DPaq7/AYBvAoAx5noRGQRw\nK8qzZPcBeL8xJrnvRgjpevJYB5Roat8Ysx7A+lS6MT/GVVZ7+b1BnpI826TuJOQRrKDT7EmVf0Z6\nOqkO8nz36ulPQiN51+00WXAzKiEkGOyACCHB6Krd8D0ACvUE9G5qdbmgrutFjDOztv9dv3mZW2e/\nnU3o82SuXc6K/smUuvuVbm1vbFGbYyk8kE9ZU9tT83/5tlropndne5z0ktrpbvRObXeWubd3mjrI\nuz3SvHtAG94/t8ayHXXutSJLCCGZwg6IEBKMrhqCjZWAUd+WcNTujrZO4siMOGXiAbDcw4LZZmc8\nUuqOB65yT1v4d8NnX9a09rj2wOlt2HNqePXSrPu1W947U00XxGOMuC93Qh3kaUtq2tXeDsZSbBCj\nB0QICQY7IEJIMLpqCPbJ+w5DoX+lX0C7neJ2Oyd8G5i8Uy0Z4F0IZm+MK//2zE0rnGbFYzBbiqqs\nWn2SsmZtj7ZlTq1JLaidRH+08hkro579xt43VNOTc/bVLPRYqW6rg7S2ZP7u1ehPW1ZNPK63P7vS\nVJ3vJDXQAyKEBIMdECEkGF01BJuYlfhqq1pip0H6To/Up0Ha5Oxk46hxqUgSkU7NThhl+8GRoscW\njw+fVVmzsEfbovz9K1f9azV9RGGqmh4sWHf9vUtsrKfv7LLDMRi1Sq5j66CD3r1a/Xm8f/WmwWbr\n3YxDD4gQEgx2QISQYHTVEKzs98XdO32iY1+vdXEvXzMae6rCNSoKnPTYZ9NGpau1qlYmbcQ7bcu6\nk0Zd4nHbxepftyaBfMqyprWnENnz+eNerl47vDhZTf/O+Eg1fdPuI6vpVxXs4sOjF49X07eesrua\nfnZqcTV9/eM2bTqoDvJujzTvHpDd+6ezve5xKz9dqjfM4hCMENIFsAMihASju4ZgEg97UFc0QRgL\nn3ySqHS+tWMu/ZnZ0sHylcs9akg8NWdnr5aqodZlr3mqmo7JG/s6Playi+F6EsSB6IQ6yEK37/2I\npd3iXtLakySDer+HnrWiTugBEUKCwQ6IEBIMMaGjjCdARE4AsAUr/hTo+22/oN5YNO1bDKYWU8UW\nLvoWg1mli9WRuNvO3WvVqCo89q7yXrWpaR3+zmNLv8cWn+1p5dOWdUAvlNPh9VLYM2WHXaKGXXev\nua+aXlZw7xV6adYugDtr22n2hhrKYUA920l1kEf79Vn5AXVc8mPnud+9NRvsPslxHVJjOof2rjcG\nm34e2HcTAKw1xmz1C9IDIoQEhB0QISQYXTULNlAwKPTWGTLGosC53fwRPRxS8gW1L0Y7l0mi0mn5\nZdEwbUxdLC6ytmh1sfAG6o7Pdp98ccCtP0lZNSUdCF7NTg0NWf2+UBQS5Vzot7KL1VDrK7vfUk1P\nqTAdRbWA7+WSfR2XL9EB7e1Qbrpj66D59tPyMd1z9qclxZSfSjwREbWWJGXVxMNx+O0plQwmvXfj\n0AMihASDHRAhJBhdNQT71qkH8NYVfpPjQcotB1V4gNM3DlfT4+q66K44wb6bknJBl6rMfnZOeYYi\nFpAxxcKspORSVlWmQeXy333Wvmp6SA2Bp+bmFyy2N0mlJ9RMlsA9JOhRYc8HevYiDf1qL9NIznWw\nTNVBxeIcmjhmlx79DBbsjdEZPQx241vo2GxZ6+UFAE/sO4CPfK+OgIIeECEkGOyACCHB6Koh2JJe\nU51lcqHvxI6U9eyjSRQ1LkFUOu3iVuzLxSVX5F1W/ZdpuVoEt7xoB1Z6Zsb1l0wPtXrUTJa3RtWN\nuQQ1qId4r6jZLt0eMbtS1oGW0HWghyUzZr5sHqQe0asHZjIuayMbltSbqa6BHhAhJBjsgAghweiq\nIdgcgHqnvhrPDyXt2qeMGnetJyqdb3jjCFGeC3mX1afnlVn7N+srvxyspiuL5tKGikiCT2dRzXx9\ndvVYNT3nqY9W2rvkOaI4RHvrX9pWIiI2W9bafGuZq3OvFnpAhJBgsAMihASjq4ZgLnxuvmdiyIsv\nalySfNsV0CSzIU2CssYXCFqmlX99/WND6kaktI1jEVGzNRcfbQPaFxyRGmtJ294xE9rU4L5s0r57\nacvq+7yQB/SACCHBYAdECAlGVw3BegAU6txP4rLqk51n1IzO+s3L3A+riIj9KnPt5hcd0zR5j0Ri\n4RbUZW1XK2Xt0xWtFzeqyhxSQ6CxSL0rUH2r+IYEi4vu+RZfe0+yveN4yhprNx3CJOFCxDReDT0g\nQkgw2AERQoLRVUOwsRIwWhuszROyYChJyTxR4zQ6ImI80p57aiHX8AyK+MI0m5u2y+dKJylrErR8\nyTEUyWUhoq5rz5DAmy/bO0aS9tazZvp3rzoUdxR8rN5q4RroARFCgsEOiBASjK4agn3yvsNQ6F/p\ndbd1GID732cj6iVZmAa3hx27Ma5c0DM3rXBIxAO1V+gVnwvsdtzTyheVvH5ywrd5J49xUj7qyjrz\nUMr2jhFfeOqug9M32joYm/Xv/StNuYPau6AHRAgJBjsgQkgwumoINjEr8dVWQDzKXWwGI8G8RJII\nefpMJeUSHxwpuqTV8bVK31SCo27zkF+kj961ydnJ5qM/NovO0fdXL8XkSXOwvWv0u9vbt3dMn7E2\nWQlw6VI9m+B3LyJ3D0hELheRORH5as31q0Rkl4iMi8gmEVmVty2EkM4i1w5IRN4J4NMAHqm5fhmA\nS6J7JwIYA7BRRPrytIcQ0lnkNgQTkSUAvgXgYgBfqLl9KYCrjTEbItmLAOwBcD6AO+pohdT6fGpG\noNfj+bUSNe5LKmqcqAh8604adcpfE8mLOnJ43Zr6sm2RV7YnKetc1rNjypcvzXgGW0W1IUmSu/G1\nsL3zae/Y9kBH+7hD3NcnTw/oZgB3GWP+WV8UkaMAHAHgnso1Y8wogM0ATs7RHkJIh5GLByQiHwew\nBsA7HLePQLmz3FNzfU90jxByiJB5ByQirwVwA4AzjTEzjeTTKY/+5TBZ4/vy7xsJpJHPU3em8m7x\nVMSOXVa6e3uss/1Hx6yspnVAjW/stEcCT87aO6JmPrNalMj2zmj/msxPetZPOsnDA1oLYAWArWIH\nigUAp4nIJQDehLKtKxH3glYCeLiu5oMbYHoG4tcWvR0yuCYTwwkh6TDj24DxR+IX5yYTPy8m43Xu\nIrIYwBtqLv8dgCcBfNkY86SI7ALwFWPM16JnhlDujC4yxsw71l5ETgCwBSv+FOj77fhNZf+AOqnz\nsfP2ukSwZoP96zuudzhPe7rtAb22Qke3arB2Q8tOJ1jnkYf8Io/8lG9diC3rYlWX285NUJfR2g+9\nTEuvnLlitW23FX32zuJea/vd+0aq6Tuf319Nv6LXd3kCkiWyke1doz9dex97l63LqcpaH5fq6eeB\nfTcBwFpjzFZ35mUy94CMMWMAfqmvicgYgP3GmCejSzcAuFJEfg3gGQBXA3gOwJ1Z20MI6VzatRI6\n5mYZY64XkUEAtwJYDuA+AO83xky3yR5CSAfQlg7IGPNex7X1ANan0TNQMCj0Gu/m3iXFlMNJT9Am\nraWkhEStORoasjt+3buj1fqkAffu4Phu5+bli0o+HkRLueo6tq8qq9aYNiCZiznlsy9SQ623Ll1U\nTX9dfWxeXrSv4OsGrPwS9ezLE1M2A2lh5QjbO3V7a/ll6vdrzDECq+6GLxkk/QrEzaiEkGCwAyKE\nBKOrdsN/69QDeOuKGpM9MaEH1Rkio2r2w7dEQXzjOjWkGFQu6N1n2WHEMhUILUyMYMtBtRP59I3D\n1fS4uh4bxeS3AR7Tc3aWZUmvXcj/56teU03r5f0H1dqfv3lmr7qTfW2yveNo+0uqDpaqzH52jm2T\nejGhn9h3AB+ZN5fthh4QISQY7IAIIcHoqiHYkl4T+xJfi76TyCVWD8z4AlT1u2cQlqtTQbVLPmNS\n5N8CuqyxkzodS+NrH/AG40oZkExHHyhUZ0XUwlD1522p2um+tOAeEhdikzh2ONar/0zqmay0lcz2\njpOgvfVQVf/u1Suvjs3eCHpAhJBgsAMihASjq4Zgc6gfN9g3U9BKgKprPUGbSo7TMaHsa6dLjphd\nNt1KMC5fgCp9eVTvSarurbIlPzBr71+89XkrqpT3qs1jr8za1n15TNWg3mCmtmqN9Fn9vuE32zt9\nWX1DOd+pP7XM1blXCz0gQkgw2AERQoLRVUMwF77v7SaBjMYXtMkrn8SIjMkqmyQBqnzDGDXqwZ+/\nzcYmnpmT+c8piydKNryGjuttlEyPMmygp/HfxqKKe6xnhqbUGCDJwtO0EWlyOanVlY/nuq+dfKQt\na9rfnVagB0QICQY7IEJIMLpqCNaD+N6heehFaupywTOhMqNmadZvXubWqaLG9avMY66pXpDXpoVp\neZd1QJW1R4WlWKoWo12ljodx/SXTdaBt6elxO/b6JNLZBFMpWuQVFYqiMhwE4vUx0UJ7az2xhYAB\nFiL6rmf1bjcqK1C/vGm8GnpAhJBgsAMihASjq4ZgYyVg1B08DkDtQkTrJI7MiFPGFyFPkyRqnJ5Z\nqNgn+rkc/PO8y6qHNzGdapj0wsR89b5i6ziWB8fsa9erwqb09c4500noV39Ktb0xLSnrIB5t0D2V\nlGs4Dk+omaEkv7U5lrVG/TzG6q0WroEeECEkGOyACCHB6Koh2CfvOwyF/pV+gVioBrfbOeHb0OKL\nkOe53KMeGFe+6ekbVwAAxlREOr1gTuuY9Rwh6bPdJ18Ut/4kZTWecZou05mbVjjVzDjs0WWFmo3q\nV+Es/uC0Z6vpf9s7WE0/vc8Grt+x217XwyhdB/EQFZ1TB2nbT8vHdKv60yEu7n+fjUyoZ6y8Xwta\nKKsmZn+dTxOlqTrfSWqgB0QICQY7IEJIMLpqCDYxK/HVVrXEjqP1HV+rj961ydnJdFHjfPtrRqIQ\nFZOxY4ATHL2LBLanlU9bVn1WlqrLgyNFl7TbHrUADsa+Xp86w8r+5w9bu3616zfV9Pc2W9f/F9s9\ni+cGfMcPh66DHNqvz8rHZ6wSzLkliYjYSnvXM2E2gX0R9IAIIcFgB0QICUZXDcHKfl/cvdPH5+rF\na5erfUr6iWtUFDhRMzatRAmMTa5EPxSL1pZ1x4/CRcwWUfJrcpBPWVYtv+6k5Pas89T71x60sr94\n4phq+vP/8eFqekjV2efeNF5NH3aYPZr5L7YNzcuzNl+vjTnWQR7tp23xfXloJfpjkrLqV/66x638\ndKneMItDMEJIF8AOiBASjO4agknNkbr1RH1hBHyubIIogTWmuNTbZz2bokwrtiQwJq3+JHWTRr9P\nx5yaZSmo4cHyxXbR2oCa9ZlWe8RcQ9xWbJx3Pcc6aKX9NL53spVon0ns8R7lnHAyuhH0gAghwWAH\nRAgJhph2RdhuARE5AcAWrPhToO+3/YJ605Bv8d8i30I230JEq3Sxiga47Vy7H0dX4ZoN5b1q47GF\niL6FYHqRnJKZzMj2EPpn3AsRP/EeK3vrp5+0qqd3VtNnX/uOanrrI4erfNTeon5PmULXQR66m3z3\ngJzev9hCxDrjrOnngX03AcBaY8xWvyA9IEJIQNgBEUKC0VWzYAMFg0JvnSFjLAqcOyTAiB4iKPnC\ngHvWJUlExMa2uENFlJSQXlA5NNS87ZoQ+ofUrJYY67Lf86h196/7nt1r9NDTx1fTjz/1qmr6VcN2\n8eGcPo65Q+sgK92ZvHs1irJ6/+LhOPzGlUoGkwnNpAdECAkGOyBCSDC6agj2rVMP4K0r/CbHA7Vb\nDqrwAKdvHK6mx9V10V1xxhODvkVv2o1dpGY5fnr2Pqf4KT+2tk8ksb2N+gcj/XefZXUvU3nq6H4j\nE3ZP0e++3g7TFh+zy+apQ5y00n5tqoNcdGdEkvdvUNkfa0P1ySNpUPon9h3AR76XzDZ6QISQYLAD\nIoQEo6uGYEt6TcytryUepFz9IM5ksqhxnoiIDVGPzSTQrSV8ZWzF9nbpX66Cz8fdd5seXqRmuFQ+\nJTVM69VtllH75VkHeddvalp4/3xtOJNgjyQQD6DfCHpAhJBgsAMihASjq4ZgcwDqnfpqPD+UVLqV\nqHFJIiJWZFrRrbe0mRxsz0r/tQ79Jc9RvnpRXcnTiJ4JmlzaL+s6bmf7aVqJiOhqP8Dfhr4j1mpJ\nc6g2PSBCSDDYARFCgtFVQzAXWa3bShsRMY0NrehOJBNIfyP8slZhJ7RfnnWcd/u1EhExCXlH68nF\nAxKR3xKR20TkRREZF5FHopg+WuYqEdkV3d8kIqvysIUQ0rlk3gGJyHIADwCYAnAOgDcD+K8AXlIy\nlwG4BMCnAZwIYAzARhHpy9oeQkjnkscQ7HIAO40xF6trv6mRuRTA1caYDQAgIhcB2APgfAB3+BQL\n5rulhVqBCF2wgrquz1eaUccIr9/sOQpYRaXrU5nFem6H/skWdGu0B6xt7zT9/ZF+XdexxaBJhhZt\nbD9NFnXQzvZr9O41o79f6U/Thq4ZrxQx6XMZgp0H4BcicoeI7BGRrSJS7YxE5CgARwC4p3LNGDMK\nYDOAk3OwhxDSoeTRAb0RwB8D2A7gbAB/DeBGEflUdP8IlPvSPTXP7YnuEUIOEfIYgvUAeMgY84Xo\n50dE5G0APgPgtlYUm8o/5Qq+ouN8q+u96tP/iArQHfuo74kap9EL6LTEqFqhZYxDf0rdqekw/ZWk\nrmt4FrT5XPR4OJXs2y81eerP891rUb+3DfUQLBJZrHqQNEOvqm1NPNOIFwA8WXPtSQD/IUrvRtnW\nlYh7QSsBPIw6/MW992JpX/w79ftWrca5x6xuxV5CSJP8YPt23Ll9OwDbAY1OTyd+Po8O6AEAtT3C\nakQfoo0xO0RkN4AzADwKACIyBOAkADfXU/zF007DscPDMU9nvN7eDEJIrpy/ejXOX13+da90QI/t\n3Yv3334QCN0VAAAUlElEQVR7oufz6IC+BuABEbkC5RmtkwBcDOA/KZkbAFwpIr8G8AyAqwE8B+DO\neooLkcE6Qt67f7Kimh5T14s9tpfS7uWEb0OLd7GWvTGhXN/3qHw1kxX9Ht15R7zLZeFYgroZj+rm\nzE3ueokHNG+cT6+4hZK0X9fVcdbvXo3ONO0HxNtQq9FRLSshN/7lHHtG2VDUm6T5sJx5B2SM+YWI\nXADgywC+AGAHgEuNMd9RMteLyCCAWwEsB3AfgPcbY5L7boSQrieXrRjGmB8B+FEDmfUA1jelX6VH\n1Amok7HTIH2nU+rTIG1ydtLzCU0faaL+ih8cKbqkrf4kutsYcKoV/d6AWY668dZL7FRNX57qxlRG\n7depddxk/QIJ3r1a/Vm92322DSsfpHXpK15ivUNTa+FmVEJIMNgBEUKC0ZW74WPf1/Sy8aJ1Qdcd\nP+p89hoVhEnUh+okQaG0/LqT6utPqzuPgFPt1F8pr64XneV1j1vZ6ZLNVZ/I2ddr87x8jdWjbWyl\n/fKsg3bVL9D43auVz0q/lq9s+2hm7Y+GHhAhJBjsgAghwejKIZgmtq4h9kneLeP7Qu87PTKJfCP9\naQNOpV3SEiqglbNuEhifZJYks/ZonFXmdZxZ+2Xw7s27nkDev26o8fW00AMihASDHRAhJBhi8g76\nmgFRONctGy+8EG8fHsZBteBw7Q9XVtPjsYWIvsVXerGWjuDk6Yv1Ajot713oWJovO+VbxGZtWaxO\n09x2rl3erptnzYaUZfXof1Tp1xx7VwZ1maQefb6/jm7VqH51nkDqOs6sDiL9ubRfu97VWvkUdbnl\ng3Y/+fJo/eKje/fi7G/fDgBrjTFb3crK0AMihASDHRAhJBhdPwsWwxOESQ8yS0pIL4IbGrLbgWM7\ngI17CqE4MOu4CozM9MxTUlC2aG0hAoYBrQW0alSXuh7jtiSYWonl6dZTrd8a+SR1nHdQr1R00rta\noyiX99UDPSBCSDDYARFCgrGwhmAK3+Ir7f4PqpmLu8/aV00v63XHMtboijuoAqGdvnEYADCuronu\n5gMFDMsqoJU0qMsk9eiNxqHSjeoX8Ndx3nWQdRuGfleBNr+vCnpAhJBgsAMihARjYQ3BUkaw0xLL\n++x17dbOeFzQom+I4rDFG/EuQMRCIMOIeo66TFKPSYZgDeu31sYOqINUdNK7WmNPLu+rB3pAhJBg\nsAMihASjK4dgrUSZu9YTIa/kOcXTe+xY7Nn59nRqxEIgu4h6rrpMUo9JhmCN6hfojDrIMyJiu97V\npPa43ldGRCSEdC3sgAghwejKIVgS0p5k6ZsR8Mn40mlsSRsRMUQ0x3ryTln9Q8rIhM3WLxA+qqC+\nn1VERK+8/iHjd7XWnrRtmBZ6QISQYLADIoQEoyuHYNrr7FF+4eSs7U/Xb17mflhFdesvuHVqkrid\nLntmUtrS57FFp/WsxYyKcrf+weUe/TrinZX/ok9+kVs+TV0mGnKodBJ5X3snquNYVMEEdeCpM28d\nR8cVF/rc9nbDuwrU1KWvrAO2bvpStHc96AERQoLBDogQEoyuHIJ5yTOCXSBbtFu9TIVkGFM3iovc\nEe9muzBCXiraGFVQR2iMyc+Vf1pSTFkZnfSuIl43yzxRLUd8wfNbgB4QISQY7IAIIcFYWEMwhXdf\nTACSLDTTzq3e67NUtdDPzrHnTcUiEybwjLOKNph3hLxmySOqYKIBR/SoHrkNFqy+UXX+l3fhacqF\niK0Q25um0v2qDn56tq0bbc7J/2Tfj8qeMu4FI4R0LeyACCHBWLBDME2Q06dTRrzzod1zPQuW1vXN\nJdpgDhHyUpFzVME0dZxyRJzZ+9EKPu2Le913fPvkWoEeECEkGOyACCHBOCSGYHnjinrXSsQ731Ao\nSVRBH3lEG3TZn0fIhlBRBZsdgulfqjxsz6qOfXoO65vz3MkeekCEkGCwAyKEBINDsAhfCAXfTEES\nmQppF5ql0Z1UTyuksT9tPWb1bCuL+VqZJXU9mrftmb2reiZOraS8SgfmV+KzamSmw8e0Aj0gQkgw\n2AERQoLBIVhErw5vkSDIu664grrebEREHfGukMQWt0Y/KWwHmre/JdvbZGPLdrroNNtT2jOr9v7d\n+MiQ256+5iJg1oMeECEkGOyACCHBOKSHYPFogO6pCHdsOKBX+dz62eqTKSPepbWllYWIDW2vySCN\n/a3Y3i4bW7XTRafZrkltj2cPWh4RGjP3gESkR0SuFpGnRWRcRH4tIlc65K4SkV2RzCYRWZW1LYSQ\nziaPIdjlAP4IwGcBvAnA5wF8XkQuqQiIyGUALgHwaQAnAhgDsFFE+uarI4QsVPIYgp0M4E5jzI+j\nn3eKyCdQ7mgqXArgamPMBgAQkYsA7AFwPoA7sjBCr5PS0epK6odx5bOeuWlFNe0PUm4pqpCE2jOd\nqGwm8oS8MB6/txVbesXnMjdpe23GKezXtntt8bnyehjgKVMWNtbamXUd51m/QA7vah178o6AmYcH\n9HMAZ4jI0QAgIscBeBeAH0U/HwXgCAD3VB4wxowC2Ixy50UIOUTIwwP6MoAhAP8qIiWUO7l1xpjv\nRPePQLlf3VPz3J7oHiHkECGPDuhjAD4B4OMAfglgDYC/FJFdxpjbcsjPoqP4TXnmMPT5UcplPThS\ndMv3e/zUKY/zuGhuvi2TgWxp1nagafsT2e6dBmuPjYntTFXHXWh7WntyiNCYRwd0PYDrjDHfi35+\nQkSOBHAFgNsA7Ea5dlYi7gWtBPBwPcVfvPdeDPX1xca6s2PvRO/iNZkZTwhJzoZfbccPn9oOAChG\n/dzo1HTi5/PogAYx/1ScOUTfm4wxO0RkN4AzADwKACIyBOAkADfXU/wXp52Gtw8P46Bay7D2hyuz\ns5wQkopzj1mNc49ZDQBYHjlmj+7di7O/fXui5/PogO4CcKWIPAfgCQAnAPgcgK8rmRsimV8DeAbA\n1QCeA3BnkgxaiTL3JR1qoMc+vO6kUaf8NVperP51a+rLa92hbWlKvkn7te3aYb/ucSs7XbItqI8E\n7uu1Nl6ubNTtnYWNtXZqmm7vNtVvHrY3Y09WETDz6IAuQblDuRnAMIBdAP46ugYAMMZcLyKDAG4F\nsBzAfQDeb4xJ7rsRQrqezDsgY8wYgD+L/tWTWw9gfdb5E0K6h67fC+Y7q8h3VG8SeZOxfCfZ0qp8\nQ/sTLGJLcr5UK3UWuo5zrd8W9WdlT1ZnhHE3PCEkGGKCHBuaDhE5AcCWH1944bxZsDV32Vmwab1L\neNq3tkKvfdBrJTKWz1N3J8snkfV9utSvYieVKYl8J9mSl7wOkBad0rvtPLuSRs+CnVOeBVtrjNnq\nVlaGHhAhJBjsgAghweiqj9C9IujtERR7bL/56gF7f1xtgS8udu8ALin3X39IKy5VMuq40Nk5z+7o\nQffQtaI/D93dIJ9Wt49OKlNa+VZ0a49A77TX693aWdbYbnu1vHhxFDhH/y4WojVdhRSrg+gBEUKC\nwQ6IEBKMrhqCjUxP48DkFPR8yj+++1mnrG9yb2bG3piesZpmpu31oSFbLQU1rEsyYVjR34rurNZY\nkM7FeH6YmLPvjX4N+tVQRx+no/WEeG1mVKDoA1HQtJHp5Bsa6AERQoLBDogQEoyuWoi4qLcXhZ54\nn9nr8Tv1MEY/8oYjF1XTr3utnUI78qjBavqO775QTU9MqAGfOJNO/a3ozjsOLwlP7P1UjX/iMnsq\naUn9bj76st2hPjmnP0KExXVUT2luDhOzswAXIhJCOhl2QISQYHTVLFjk1qWmb8C6uMedbI80OeWU\nV9nra+xqwWf3j1XT3//7vVaR7q6VF+zSn5VusvAZ7LHToU+N2qHWiUN2OPaeoWXV9N/vse/NTJeP\n1+kBEUKCwQ6IEBKMrhqCNUJPJBVUyaYnrZvaq6asRg/aId3HPvJINb3r+Umr0zM71Uh/Wt1dMBlJ\ncmJizm6y2jk1VU2vUOnX9PdX0xe+xh6fd9sLdlZVexO1p0J0KvSACCHBYAdECAnGghqC6VGMb0jT\n32/73KVL3fuySnPuh7XORvrT6u7yyQySEv2XXzf9hStthM/9atb3Y8PD1fSSXvtufXf37mp6ugvH\n8fSACCHBYAdECAnGghqCaXzhLR544KVq+pOffE01/cMfrq2mL7/8V9X0979vg24nCZ9R0d+K7lK3\nTGGQptGDJR1q46NqCDajXrIbdu6spqfUXjA97Oq+ARg9IEJIQNgBEUKC0VXhOJp/3qZ1cdeoPVrH\nH2/33Xz3u3ZmYXzcjoeSnDBZuZ6VbtIavtM8Q0e08EUvPHn5cuf1nx88mJ8x+cFwHISQzoUdECEk\nGIfEEIwcWizE4JK+IVuHl49DMEJI58IOiBASjAW7EDEJOpi8niHJYiFgnrrJfHzDrn/3ZhuZ8oSj\n7QzT39y1o5r2bc8LQZcOtZqGHhAhJBjsgAghwTikh2B5LkYLvdDtUMAXrfK9x9uDB8484fBqesuv\nRqrpay9+SzX937/xZDU9U1J7qwKMexbqUMsHPSBCSDDYARFCgnFID8FId9OjZxfV2GXt0fYMrX0H\nbWD379+3q5r+7IeOqqb7ivbv8PQspynbCT0gQkgw2AERQoLBvWCka4mF2lDXBwfsl4VLzrdDrdce\nvqia/s5Pn6+m7398v1NnF/xqdDrcC0YI6VzYARFCgsEhGFkQ+PaC6ZmyVy3tq6b3j07nbhPhEIwQ\n0sGwAyKEBIMLEcmCwPchQYfa0MMuznZ1Bqk9IBE5VUT+UUSeF5E5EfmQQ+YqEdklIuMisklEVtXc\n7xeRm0XkRRF5WUT+XkSGa/UQQhY2zQzBFgPYBuCzcPzhEZHLAFwC4NMATgQwBmCjiPQpsRsAfBDA\n7wI4DcBvAfi/TdhCCOlmjDFN/wMwB+BDNdd2Afic+nkIwASAj6qfpwBcoGRWR7pO9ORzAsqdHf/x\nH/91z78TGvUhmX6EFpGjABwB4J7KNWPMKIDNAE6OLr0D5W9PWmY7gJ1KhhByCJD1LNgRKPd8e2qu\n74nuAcBKANNRx+STIYQcAnAanhASjKw7oN0oL0pdWXN9ZXSvItMnIkN1ZAghhwCZdkDGmB0odyJn\nVK5FHc1JAH4eXdoCYLZGZjWA1wP4f1naQwjpbFIvRBSRxQBWwW6/eaOIHAfggDHmWZSn2K8UkV8D\neAbA1QCeA3AnUP4oLSJ/C+CrIvISgJcB3AjgAWPMQy2WhxDSTTQx9f5ulKfMSzX//peSWY/ydPw4\ngI0AVtXo6AdwE4AXUe6AvgdguE6enIbnP/7rvn8Np+G5G54QkhfcDU8I6VzYARFCgsEOiBASDHZA\nhJBgsAMihASDHRAhJBjsgAghweiWDmggtAGEkNQ0/L3tlg7oyNAGEEJSc2QjgW5ZCf1qAOegvLds\nMqw1hJAGDKDc+Ww0xuyvJ9gVHRAhZGHSLUMwQsgChB0QISQY7IAIIcFgB0QICUbXdEAi8iciskNE\nJkTkQRF5Z8b6Wz7xtYk8rxCRh0RkVET2iMg/iMgxeeYrIp8RkUdEZCT693MReV9e+dWx4/Konr+a\nV94i8sUoD/3vl3nlV6P3t0Tktuj03/Gozk/IM+/o96O2vHMiclNeebZMKwcTtusfgI+hPP1+EYA3\nAbgVwAEAh2eYx/sAXAXgwyhHeKw9cPGyKM9zAbwNwA8A/BuAvhby/BGATwF4M4BjAWxAeanBorzy\nRflE2vcB+B2UQ+teg/JBkW/Oq5wOG94J4GkADwP4ao5l/SKARwGsADAc/TsszzaN9C4HsAPA1wGs\nBfAGAGcCOCrn9+nVqpzDKMddLwE4tV1tm9rmUBmnrNgHAfyl+llQjjP9+ZzyS33ia0b5Hh7lfUqb\n890P4A/akR+AJQC2A3gvgJ/WdECZ5h11QFvr3M+lrAC+DOBfGsi0o11vAPCrduaZ9l/HD8FEpIjy\nXxF9kqoBcDfadJJqwhNfs2A5yrF0D7QjXxHpEZGPAxgE8PM2lfNmAHcZY/65xpa88j46Glb/m4h8\nS0Rel3N+AHAegF+IyB3R0HqriFxcudmOeo5+b34PwN+2K89m6PgOCGWvoID6p63mTZITX1tCRATl\nv1j3G2Mq3ylyyVdE3iYiL6M89LoFwAWmfDx2ruWMOrs1AK5w3M4j7wcB/D7Kq+g/A+AoAPdGJ7vk\nWdY3AvhjlD29swH8NYAbReRT0f3c3ycAFwBYBuB/tzHP1KQ+lofkxi0A3gLgXW3I618BHIfyC/oR\nAN8UkdPyzFBEXotyB3umMWYmz7wqGGM2qh8fF5GHAPwGwEdRroO86AHwkDHmC9HPj4jI21DuBG/L\nMV/NHwL4J2NMRx/22Q0e0Isof0ird9pq3iQ58bVpROSvAHwAwOnGmBfyztcYM2uMedoY87AxZh2A\nRwBcmld+EWtR/hi8VURmRGQG5SOeLhWRaZT/EudWxwBgjBkB8CuUP77nWdYXADxZc+1JlA/fRM55\nQ0Rej/JH7/+pLueaZ7N0fAcU/bXcgvhJqhL9/HPfcxnbsAONT3xtiqjz+TCA9xhjdrYr3xp6APTn\nnN/dKM/0rUHZ+zoOwC8AfAvAccaYp3PMu6JvCcqdz66cy/oAgNU111aj7H21o13/EOUO/UeVC218\nl9IR6ut3yq/5H0X5kEM9Db8fwIoM81iM8i/FGpRnov5L9PProvufj/I8D+VfpB8AeAqtTZveAuAl\nAKei/Jeo8m9AyWSaL4Bro/zegPJU7HUoH5X93rzKWceW2lmwrMv6FQCnRWX99wA2ofyL+eo8ywrg\nHSh/X7sC5eUOn0D5AM6P51VWpVdQXsrxJce9trVtYntDZdxExX42qtgJlM+Qf0fG+t+NFk98bSJP\nV34lABfVyGWWL8prU56O6nE3gJ9UOp+8ylnHln/WHVAOZb0d5eUaEwB2Avg21FqcPMuK8pD60Ujv\nEwD+0CGTed4AzoreIaeudrVt0n8Mx0EICUbHfwMihCxc2AERQoLBDogQEgx2QISQYLADIoQEgx0Q\nISQY7IAIIcFgB0QICQY7IEJIMNgBEUKCwQ6IEBIMdkCEkGD8fyGuTDcxu66zAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10c6d7390>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from env import Atari\n",
    "\n",
    "#creating a game\n",
    "atari = Atari(GAME,image_size=IMAGE_SIZE) \n",
    "\n",
    "action_names = np.array(atari.get_action_meanings())\n",
    "\n",
    "obs = atari.step(0)[0]\n",
    "\n",
    "plt.imshow(obs,interpolation='none')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'GAME_TITLE' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-e4f07a0a61bc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mgym\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mgym\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mmake\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0matari\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgym\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGAME_TITLE\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0matari\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0matari\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrender\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'rgb_array'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'GAME_TITLE' is not defined"
     ]
    }
   ],
   "source": [
    "import gym\n",
    "from gym import make\n",
    "atari = gym.make(GAME_TITLE)\n",
    "atari.reset()\n",
    "plt.imshow(atari.render('rgb_array'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import lasagne\n",
    "from lasagne.layers import InputLayer, DimshuffleLayer\n",
    "\n",
    "#image observation at current tick goes here, shape = (sample_i,x,y,color)\n",
    "observation_layer = InputLayer((None,IMAGE_W,IMAGE_H,3))\n",
    "\n",
    "#reshape to [sample, color, x, y] to allow for convolutional layers to work correctly\n",
    "observation_reshape = DimshuffleLayer(observation_layer,(0,3,1,2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from lasagne.layers import Conv2DLayer,Pool2DLayer,DenseLayer,batch_norm,dropout\n",
    "\n",
    "#main neural network body\n",
    "conv0 = Conv2DLayer(observation_reshape,16,filter_size=(8,8),stride=(4,4),name='conv0')\n",
    "\n",
    "conv1 = Conv2DLayer(batch_norm(conv0),32,filter_size=(4,4),stride=(2,2),name='conv1')\n",
    "\n",
    "conv2 = Conv2DLayer(batch_norm(conv1),64,filter_size=(4,4),stride=(2,2),name='conv2')\n",
    "\n",
    "dense0 = DenseLayer(batch_norm(conv2),256,name='dense',nonlinearity = lasagne.nonlinearities.tanh)\n",
    "\n",
    "#please set this to your last layer for convenience\n",
    "last_layer = dense0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#a layer that predicts Qvalues\n",
    "qvalues_layer = DenseLayer(last_layer,\n",
    "                   num_units = atari.action_space.n,\n",
    "                   nonlinearity=lasagne.nonlinearities.linear,\n",
    "                   name=\"q-evaluator layer\")\n",
    "\n",
    "#To pick actions, we use an epsilon-greedy resolver (epsilon is a property)\n",
    "from agentnet.resolver import EpsilonGreedyResolver\n",
    "action_layer = EpsilonGreedyResolver(qvalues_layer,name=\"e-greedy action picker\")\n",
    "\n",
    "action_layer.epsilon.set_value(np.float32(0.1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from agentnet.agent import Agent\n",
    "#all together\n",
    "agent = Agent(observation_layers=observation_layer,\n",
    "              policy_estimators=qvalues_layer,\n",
    "              action_layers=action_layer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[conv0.W,\n",
       " conv0_bn.beta,\n",
       " conv0_bn.gamma,\n",
       " conv1.W,\n",
       " conv1_bn.beta,\n",
       " conv1_bn.gamma,\n",
       " conv2.W,\n",
       " conv2_bn.beta,\n",
       " conv2_bn.gamma,\n",
       " dense.W,\n",
       " dense.b,\n",
       " q-evaluator layer.W,\n",
       " q-evaluator layer.b]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Since it's a single lasagne network, one can get it's weights, output, etc\n",
    "weights = lasagne.layers.get_all_params(action_layer,trainable=True)\n",
    "weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR (theano.gof.opt): Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "[2016-10-27 22:12:02,368] Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "ERROR (theano.gof.opt): node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(2, 2), filter_flip=True, imshp=(None, 32, 11, 8), kshp=(64, 32, 4, 4), filter_dilation=(1, 1)}(Elemwise{mul,no_inplace}.0, conv2.W)\n",
      "[2016-10-27 22:12:02,370] node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(2, 2), filter_flip=True, imshp=(None, 32, 11, 8), kshp=(64, 32, 4, 4), filter_dilation=(1, 1)}(Elemwise{mul,no_inplace}.0, conv2.W)\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "[2016-10-27 22:12:02,372] TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "[2016-10-27 22:12:02,375] Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "[2016-10-27 22:12:02,378] Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "ERROR (theano.gof.opt): node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(2, 2), filter_flip=True, imshp=(None, 32, 11, 8), kshp=(64, 32, 4, 4), filter_dilation=(1, 1)}(Elemwise{mul,no_inplace}.0, conv2.W)\n",
      "[2016-10-27 22:12:02,381] node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(2, 2), filter_flip=True, imshp=(None, 32, 11, 8), kshp=(64, 32, 4, 4), filter_dilation=(1, 1)}(Elemwise{mul,no_inplace}.0, conv2.W)\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "[2016-10-27 22:12:02,385] TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "[2016-10-27 22:12:02,388] Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "[2016-10-27 22:12:02,390] Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "ERROR (theano.gof.opt): node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(2, 2), filter_flip=True, imshp=(None, 16, 25, 19), kshp=(32, 16, 4, 4), filter_dilation=(1, 1)}(Elemwise{mul,no_inplace}.0, conv1.W)\n",
      "[2016-10-27 22:12:02,392] node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(2, 2), filter_flip=True, imshp=(None, 16, 25, 19), kshp=(32, 16, 4, 4), filter_dilation=(1, 1)}(Elemwise{mul,no_inplace}.0, conv1.W)\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "[2016-10-27 22:12:02,393] TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "[2016-10-27 22:12:02,396] Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "[2016-10-27 22:12:02,399] Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "ERROR (theano.gof.opt): node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(2, 2), filter_flip=True, imshp=(None, 16, 25, 19), kshp=(32, 16, 4, 4), filter_dilation=(1, 1)}(Elemwise{mul,no_inplace}.0, conv1.W)\n",
      "[2016-10-27 22:12:02,401] node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(2, 2), filter_flip=True, imshp=(None, 16, 25, 19), kshp=(32, 16, 4, 4), filter_dilation=(1, 1)}(Elemwise{mul,no_inplace}.0, conv1.W)\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "[2016-10-27 22:12:02,402] TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "[2016-10-27 22:12:02,404] Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "[2016-10-27 22:12:02,406] Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "ERROR (theano.gof.opt): node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(4, 4), filter_flip=True, imshp=(None, 3, 105, 80), kshp=(16, 3, 8, 8), filter_dilation=(1, 1)}(InplaceDimShuffle{0,3,1,2}.0, conv0.W)\n",
      "[2016-10-27 22:12:02,407] node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(4, 4), filter_flip=True, imshp=(None, 3, 105, 80), kshp=(16, 3, 8, 8), filter_dilation=(1, 1)}(InplaceDimShuffle{0,3,1,2}.0, conv0.W)\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "[2016-10-27 22:12:02,409] TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "[2016-10-27 22:12:02,411] Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "[2016-10-27 22:12:02,413] Optimization failure due to: LocalOptGroup(local_abstractconv_gemm,local_abstractconv_gradweight_gemm,local_abstractconv_gradinputs_gemm,local_abstractconv3d_gemm,local_abstractconv3d_gradweight_gemm,local_abstractconv3d_gradinputs_gemm,local_conv2d_cpu,local_conv2d_gradweight_cpu,local_conv2d_gradinputs_cpu,local_conv3d_cpu,local_conv3d_gradweight_cpu,local_conv3d_gradinputs_cpu)\n",
      "ERROR (theano.gof.opt): node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(4, 4), filter_flip=True, imshp=(None, 3, 105, 80), kshp=(16, 3, 8, 8), filter_dilation=(1, 1)}(InplaceDimShuffle{0,3,1,2}.0, conv0.W)\n",
      "[2016-10-27 22:12:02,414] node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(4, 4), filter_flip=True, imshp=(None, 3, 105, 80), kshp=(16, 3, 8, 8), filter_dilation=(1, 1)}(InplaceDimShuffle{0,3,1,2}.0, conv0.W)\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "[2016-10-27 22:12:02,417] TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "[2016-10-27 22:12:02,419] Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1135, in _unify_values\n",
      "    sectiondict = self._sections[section]\n",
      "KeyError: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 166, in fetch_val_for_key\n",
      "    return theano_cfg.get(section, option)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 778, in get\n",
      "    d = self._unify_values(section, vars)\n",
      "  File \"//anaconda/lib/python3.5/configparser.py\", line 1138, in _unify_values\n",
      "    raise NoSectionError(section)\n",
      "configparser.NoSectionError: No section: 'blas'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 324, in __get__\n",
      "    delete_key=delete_key)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 170, in fetch_val_for_key\n",
      "    raise KeyError(key)\n",
      "KeyError: 'blas.ldflags'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1309, in transform\n",
      "    new_repl = opt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 75, in local_abstractconv_gemm\n",
      "    if theano.config.cxx == \"\" or not theano.config.blas.ldflags:\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configparser.py\", line 328, in __get__\n",
      "    val_str = self.default()\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\", line 1258, in default_blas_ldflags\n",
      "    lib_path = blas_info.get('library_dirs', [])[0]\n",
      "IndexError: list index out of range\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: local_abstractconv_check\n",
      "[2016-10-27 22:12:02,422] Optimization failure due to: local_abstractconv_check\n",
      "ERROR (theano.gof.opt): node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(4, 4), filter_flip=True, imshp=(None, 3, 105, 80), kshp=(16, 3, 8, 8), filter_dilation=(1, 1)}(InplaceDimShuffle{0,3,1,2}.0, conv0.W)\n",
      "[2016-10-27 22:12:02,424] node: AbstractConv2d{convdim=2, border_mode='valid', subsample=(4, 4), filter_flip=True, imshp=(None, 3, 105, 80), kshp=(16, 3, 8, 8), filter_dilation=(1, 1)}(InplaceDimShuffle{0,3,1,2}.0, conv0.W)\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "[2016-10-27 22:12:02,426] TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 603, in local_abstractconv_check\n",
      "    node.op.__class__.__name__)\n",
      "AssertionError: AbstractConv2d Theano optimization failed: there is no implementation available supporting the requested options. Did you exclude both \"conv_dnn\" and \"conv_gemm\" from the optimizer? If on GPU, is cuDNN available and does the GPU support it? If on CPU, do you have a BLAS library installed Theano can link against?\n",
      "\n",
      "[2016-10-27 22:12:02,429] Traceback (most recent call last):\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\", line 1922, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\", line 603, in local_abstractconv_check\n",
      "    node.op.__class__.__name__)\n",
      "AssertionError: AbstractConv2d Theano optimization failed: there is no implementation available supporting the requested options. Did you exclude both \"conv_dnn\" and \"conv_gemm\" from the optimizer? If on GPU, is cuDNN available and does the GPU support it? If on CPU, do you have a BLAS library installed Theano can link against?\n",
      "\n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "AbstractConv2d Theano optimization failed: there is no implementation available supporting the requested options. Did you exclude both \"conv_dnn\" and \"conv_gemm\" from the optimizer? If on GPU, is cuDNN available and does the GPU support it? If on CPU, do you have a BLAS library installed Theano can link against?",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-2021210f8c3c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mpool\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mAtariGamePool\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mpool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mAtariGamePool\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0magent\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mGAME\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mN_AGENTS\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mimage_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mIMAGE_SIZE\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmax_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10000\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/luba_yaronskaya/Documents/7semestr/ml/dqn_binder/pool.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, agent, game_title, n_games, max_size, **kwargs)\u001b[0m\n\u001b[1;32m     38\u001b[0m         \u001b[0;31m#save agent\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0magent\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0magent\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0magent_step\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0magent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_react_function\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m         \u001b[0;31m# Create experience replay environment\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/agentnet/agent/mdp_agent.py\u001b[0m in \u001b[0;36mget_react_function\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    512\u001b[0m         \u001b[0;31m# compile\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    513\u001b[0m         applier_fun = theano.function(applier_observations + list(applier_memories.values()),\n\u001b[0;32m--> 514\u001b[0;31m                                       applier_actions + applier_new_states)\n\u001b[0m\u001b[1;32m    515\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    516\u001b[0m         \u001b[0;31m# return\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/compile/function.py\u001b[0m in \u001b[0;36mfunction\u001b[0;34m(inputs, outputs, mode, updates, givens, no_default_updates, accept_inplace, name, rebuild_strict, allow_input_downcast, profile, on_unused_input)\u001b[0m\n\u001b[1;32m    324\u001b[0m                    \u001b[0mon_unused_input\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mon_unused_input\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m                    \u001b[0mprofile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mprofile\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m                    output_keys=output_keys)\n\u001b[0m\u001b[1;32m    327\u001b[0m     \u001b[0;31m# We need to add the flag check_aliased inputs if we have any mutable or\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m     \u001b[0;31m# borrowed used defined inputs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/compile/pfunc.py\u001b[0m in \u001b[0;36mpfunc\u001b[0;34m(params, outputs, mode, updates, givens, no_default_updates, accept_inplace, name, rebuild_strict, allow_input_downcast, profile, on_unused_input, output_keys)\u001b[0m\n\u001b[1;32m    484\u001b[0m                          \u001b[0maccept_inplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maccept_inplace\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    485\u001b[0m                          \u001b[0mprofile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mprofile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mon_unused_input\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mon_unused_input\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 486\u001b[0;31m                          output_keys=output_keys)\n\u001b[0m\u001b[1;32m    487\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/compile/function_module.py\u001b[0m in \u001b[0;36morig_function\u001b[0;34m(inputs, outputs, mode, accept_inplace, name, profile, on_unused_input, output_keys)\u001b[0m\n\u001b[1;32m   1774\u001b[0m                    \u001b[0mprofile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mprofile\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1775\u001b[0m                    \u001b[0mon_unused_input\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mon_unused_input\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1776\u001b[0;31m                    \u001b[0moutput_keys\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0moutput_keys\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1777\u001b[0m             defaults)\n\u001b[1;32m   1778\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/compile/function_module.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, inputs, outputs, mode, accept_inplace, function_builder, profile, on_unused_input, fgraph, output_keys)\u001b[0m\n\u001b[1;32m   1456\u001b[0m                         optimizer, inputs, outputs)\n\u001b[1;32m   1457\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1458\u001b[0;31m                     \u001b[0moptimizer_profile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0moptimizer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfgraph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1459\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1460\u001b[0m                 \u001b[0mend_optimizer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, fgraph)\u001b[0m\n\u001b[1;32m     96\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     97\u001b[0m         \"\"\"\n\u001b[0;32m---> 98\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptimize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfgraph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     99\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    100\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0madd_requirements\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfgraph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36moptimize\u001b[0;34m(self, fgraph, *args, **kwargs)\u001b[0m\n\u001b[1;32m     85\u001b[0m             \u001b[0morig\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtheano\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtensor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasic\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconstant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m             \u001b[0mtheano\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtensor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasic\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconstant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 87\u001b[0;31m             \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfgraph\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     88\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m             \u001b[0mtheano\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtensor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasic\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconstant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0morig\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, fgraph)\u001b[0m\n\u001b[1;32m    233\u001b[0m                 \u001b[0mnb_nodes_before\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfgraph\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_nodes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    234\u001b[0m                 \u001b[0mt0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 235\u001b[0;31m                 \u001b[0msub_prof\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0moptimizer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptimize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfgraph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    236\u001b[0m                 \u001b[0ml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mt0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    237\u001b[0m                 \u001b[0msub_profs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msub_prof\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36moptimize\u001b[0;34m(self, fgraph, *args, **kwargs)\u001b[0m\n\u001b[1;32m     85\u001b[0m             \u001b[0morig\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtheano\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtensor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasic\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconstant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m             \u001b[0mtheano\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtensor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasic\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconstant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 87\u001b[0;31m             \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfgraph\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     88\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m             \u001b[0mtheano\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtensor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasic\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconstant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0morig\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, fgraph, start_from)\u001b[0m\n\u001b[1;32m   2033\u001b[0m                     \u001b[0mnode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpopleft\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2034\u001b[0m                 \u001b[0mcurrent_node\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2035\u001b[0;31m                 \u001b[0mnb\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprocess_node\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfgraph\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2036\u001b[0m             \u001b[0mloop_t\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mt0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2037\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36mprocess_node\u001b[0;34m(self, fgraph, node, lopt)\u001b[0m\n\u001b[1;32m   1925\u001b[0m                 self.failure_callback(e, self,\n\u001b[1;32m   1926\u001b[0m                                       \u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutputs\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1927\u001b[0;31m                                       lopt, node)\n\u001b[0m\u001b[1;32m   1928\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1929\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36mwarn_inplace\u001b[0;34m(exc, nav, repl_pairs, local_opt, node)\u001b[0m\n\u001b[1;32m   1819\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mInconsistencyError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1820\u001b[0m             \u001b[0;32mreturn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1821\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mNavigatorOptimizer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwarn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnav\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrepl_pairs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlocal_opt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1822\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1823\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36mwarn\u001b[0;34m(exc, nav, repl_pairs, local_opt, node)\u001b[0m\n\u001b[1;32m   1805\u001b[0m             \u001b[0;31m# We always crash on AssertionError because something may be\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1806\u001b[0m             \u001b[0;31m# seriously wrong if such an exception is raised.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1807\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1808\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1809\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/gof/opt.py\u001b[0m in \u001b[0;36mprocess_node\u001b[0;34m(self, fgraph, node, lopt)\u001b[0m\n\u001b[1;32m   1920\u001b[0m         \u001b[0mlopt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlopt\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocal_opt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1921\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1922\u001b[0;31m             \u001b[0mreplacements\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlopt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtransform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1923\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1924\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfailure_callback\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/tensor/nnet/opt.py\u001b[0m in \u001b[0;36mlocal_abstractconv_check\u001b[0;34m(node)\u001b[0m\n\u001b[1;32m    601\u001b[0m             \u001b[0;34m'is cuDNN available and does the GPU support it? If on CPU, '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    602\u001b[0m             \u001b[0;34m'do you have a BLAS library installed Theano can link against?'\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 603\u001b[0;31m             node.op.__class__.__name__)\n\u001b[0m\u001b[1;32m    604\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    605\u001b[0m optdb.register('AbstractConvCheck',\n",
      "\u001b[0;31mAssertionError\u001b[0m: AbstractConv2d Theano optimization failed: there is no implementation available supporting the requested options. Did you exclude both \"conv_dnn\" and \"conv_gemm\" from the optimizer? If on GPU, is cuDNN available and does the GPU support it? If on CPU, do you have a BLAS library installed Theano can link against?"
     ]
    }
   ],
   "source": [
    "from pool import AtariGamePool\n",
    "\n",
    "pool = AtariGamePool(agent,GAME, N_AGENTS,image_size=IMAGE_SIZE,max_size=10000)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pool' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-004e2c8e4d96>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpool\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'pool' is not defined"
     ]
    }
   ],
   "source": [
    "pool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import theano"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m//anaconda/lib/python3.5/configparser.py\u001b[0m in \u001b[0;36m_unify_values\u001b[0;34m(self, section, vars)\u001b[0m\n\u001b[1;32m   1134\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1135\u001b[0;31m             \u001b[0msectiondict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sections\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msection\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1136\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'blas'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mNoSectionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/configparser.py\u001b[0m in \u001b[0;36mfetch_val_for_key\u001b[0;34m(key, delete_key)\u001b[0m\n\u001b[1;32m    165\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 166\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mtheano_cfg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msection\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moption\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    167\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mConfigParser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mInterpolationError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/configparser.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, section, option, raw, vars, fallback)\u001b[0m\n\u001b[1;32m    777\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 778\u001b[0;31m             \u001b[0md\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_unify_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msection\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    779\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mNoSectionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/configparser.py\u001b[0m in \u001b[0;36m_unify_values\u001b[0;34m(self, section, vars)\u001b[0m\n\u001b[1;32m   1137\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0msection\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefault_section\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1138\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mNoSectionError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msection\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1139\u001b[0m         \u001b[0;31m# Update with the entry specific variables\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNoSectionError\u001b[0m: No section: 'blas'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/configparser.py\u001b[0m in \u001b[0;36m__get__\u001b[0;34m(self, cls, type_, delete_key)\u001b[0m\n\u001b[1;32m    323\u001b[0m                 val_str = fetch_val_for_key(self.fullname,\n\u001b[0;32m--> 324\u001b[0;31m                                             delete_key=delete_key)\n\u001b[0m\u001b[1;32m    325\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_default\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/configparser.py\u001b[0m in \u001b[0;36mfetch_val_for_key\u001b[0;34m(key, delete_key)\u001b[0m\n\u001b[1;32m    169\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mConfigParser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mNoOptionError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mConfigParser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mNoSectionError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 170\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    171\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'blas.ldflags'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-6e162f9c89ed>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtheano\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/configparser.py\u001b[0m in \u001b[0;36m__str__\u001b[0;34m(self, print_doc)\u001b[0m\n\u001b[1;32m    202\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__str__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprint_doc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m         \u001b[0msio\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mStringIO\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 204\u001b[0;31m         \u001b[0m_config_print\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprint_doc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mprint_doc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    205\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0msio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetvalue\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    206\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/configparser.py\u001b[0m in \u001b[0;36m_config_print\u001b[0;34m(thing, buf, print_doc)\u001b[0m\n\u001b[1;32m    178\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mprint_doc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    179\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"    Doc: \"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdoc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbuf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 180\u001b[0;31m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"    Value: \"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__get__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbuf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    181\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbuf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    182\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/configparser.py\u001b[0m in \u001b[0;36m__get__\u001b[0;34m(self, cls, type_, delete_key)\u001b[0m\n\u001b[1;32m    326\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefault\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m                     \u001b[0mval_str\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefault\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    329\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m                     \u001b[0mval_str\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefault\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python3.5/site-packages/theano/configdefaults.py\u001b[0m in \u001b[0;36mdefault_blas_ldflags\u001b[0;34m()\u001b[0m\n\u001b[1;32m   1256\u001b[0m                     \u001b[0mflags\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'-L\"%s\"'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mlib_path\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1257\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1258\u001b[0;31m                     \u001b[0mlib_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblas_info\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'library_dirs'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1259\u001b[0m                     \u001b[0mflags\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'-L%s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mlib_path\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1260\u001b[0m                 flags += ['-l%s' % l for l in [\"mkl_core\",\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "print(theano.config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
